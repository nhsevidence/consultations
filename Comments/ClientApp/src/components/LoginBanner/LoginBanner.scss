@import "~@nice-digital/nds-input/scss/input.scss";

.sign-in-banner {
  margin-top: 0.5rem !important;
}

.panel-white {
  background-color: #fff;
}

.display--inline {
  display: inline;
}

.no-margin {
  margin: 0;
}
