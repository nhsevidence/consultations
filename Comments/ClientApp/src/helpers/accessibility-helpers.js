/**
 * Pull focus to an element, principally for accessibility
 * @param {string} selection - The ***ID ONLY*** If it's not an interactive element, the target will need an attribute of "tabIndex={-1}"
 */
// todo: needs test
export const pullFocusByQuerySelector = selection => {
	const el = document.querySelector(selection);
	if (el.getAttribute("tabindex") === null) {
		el.setAttribute("tabindex", "-1");
	}
	el.focus();
	console.log(`Pulling focus by querySelector to ${selection}`);
};

export const pullFocusById = selection => {
	selection = selection.replace("#", ""); // make sure there's no # being passed into here
	const el = document.getElementById(selection);
	if (el.getAttribute("tabindex") === null) {
		el.setAttribute("tabindex", "-1");
	}
	el.focus();
	console.log(`Pulling focus by ID to ${selection}`);
};
