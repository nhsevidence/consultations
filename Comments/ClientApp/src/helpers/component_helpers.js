// @flow

import React from "react";
import { HashLink } from "react-router-hash-link";

export const HashLinkTop = (
	label: string,
	destination: string,
	behavior: string = "instant",
	block: string = "start") => {
	return (
		<HashLink
			to={destination}
			scroll={el => el.scrollIntoView({ behavior, block })}>
			{label}
		</HashLink>
	);
};
