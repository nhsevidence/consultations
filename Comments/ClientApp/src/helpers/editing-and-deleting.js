// @flow

import { load } from "../data/loader";
import { tagManager } from "./tag-manager";

export function saveCommentHandler(event: Event, comment: CommentType, self: any) {
	event.preventDefault();

	const originalId = comment.commentId;
	const isANewComment = comment.commentId < 0;
	const method = isANewComment ? "POST" : "PUT";
	const urlParameters = isANewComment ? [] : [comment.commentId];
	const endpointName = isANewComment ? "newcomment" : "editcomment";
	let error = "";

	load(endpointName, undefined, urlParameters, {}, method, comment, true)
		.then(res => {
			if (res.status === 201 || res.status === 200) {
				const index = self.state.comments
					.map(function(comment) {
						return comment.commentId;
					})
					.indexOf(comment.commentId);
				const comments = self.state.comments;
				comments[index] = res.data;
				self.setState({
					comments,
					error,
				});
				tagManager({
					event: "generic",
					category: "Consultation comments page",
					action: "Clicked",
					label: "Comment saved button",
				});
				self.updateUnsavedIds(`${originalId}c`, false);
				if (typeof self.issueA11yMessage === "function") {
					self.issueA11yMessage("Comment saved");
				}
				if (typeof self.validationHander === "function") {
					self.validationHander();
				}
			}
		})
		.catch(err => {
			console.log(err);
			if (err.response) {
				error = "save";
				self.setState({
					error,
				});
				if (typeof self.issueA11yMessage === "function") {
					self.issueA11yMessage("There was a problem saving this comment");
				}
			}
		});
}

export function deleteCommentHandler(event: Event, commentId: number, self: any) {
	event.preventDefault();
	if (commentId < 0) {
		removeCommentFromState(commentId, self);
	} else {
		load("editcomment", undefined, [commentId], {}, "DELETE")
			.then(res => {
				if (res.status === 200) {
					removeCommentFromState(commentId, self);
				}
			})
			.catch(err => {
				console.log(err);
				if (err.response) {
					const error = "delete";
					self.setState({
						error,
					});
					if (typeof self.issueA11yMessage === "function") {
						self.issueA11yMessage("There was a problem deleting this comment");
					}
				}
			});
	}
}

export function saveAnswerHandler(event: Event, answer: AnswerType, questionId: number, self: any) {
	event.preventDefault();
	const isANewAnswer = answer.answerId < 0;
	const method = isANewAnswer ? "POST" : "PUT";
	const urlParameters = isANewAnswer ? [] : [answer.answerId];
	const endpointName = isANewAnswer ? "newanswer" : "editanswer";

	load(endpointName, undefined, urlParameters, {}, method, answer, true)
		.then(res => {
			if (res.status === 201 || res.status === 200) {
				const questionIndex = self.state.questions
					.map(function(question) {
						return question.questionId;
					})
					.indexOf(answer.questionId);
				const questions = self.state.questions;

				if (questions[questionIndex].answers === null || questions[questionIndex].answers.length < 1){
					questions[questionIndex].answers = [res.data];
				} else{
					const answerIndex = questions[questionIndex].answers
						.map(function(answer) {
							return answer.answerId;
						}).indexOf(answer.answerId);

					const answers = questions[questionIndex].answers;
					answers[answerIndex] = res.data;
					questions[questionIndex].answers = answers;
				}
				self.setState({
					questions,
				});
				tagManager({
					event: "generic",
					category: "Consultation comments page",
					action: "Clicked",
					label: "Save answer button",
				});
				self.updateUnsavedIds(`${questionId}q`, false);
				if (typeof self.issueA11yMessage === "function") {
					self.issueA11yMessage("Answer saved");
				}
				if (typeof self.validationHander === "function") {
					self.validationHander();
				}
			}
		})
		.catch(err => {
			console.log(err);
			if (typeof self.issueA11yMessage === "function") {
				self.issueA11yMessage("There was a problem saving this answer");
			}
			if (err.response) alert(err.response.statusText);
		});
}

export function deleteAnswerHandler(event: Event, questionId: number, answerId: number, self: any) {
	event.preventDefault();
	if (answerId < 0) {
		removeAnswerFromState(questionId, answerId, self);
	} else {
		load("editanswer", undefined, [answerId], {}, "DELETE")
			.then(res => {
				if (res.status === 200) {
					removeAnswerFromState(questionId, answerId, self);
				}
			})
			.catch(err => {
				console.log(err);
				if (err.response) alert(err.response.statusText);
				if (typeof self.issueA11yMessage === "function") {
					self.issueA11yMessage("There was a problem deleting this answer");
				}
			});
	}
}

export function saveQuestionHandler(event: Event, question: QuestionType, self: any) {
	console.log("Called saveQuestionHandler in editing-and-deleting.js");
	console.log(event, question, self);
}

export function deleteQuestionHandler(event: Event, questionId: number, self: any) {
	console.log("Called deleteQuestionHandler in editing-and-deleting.js");
	console.log(event, questionId, self);

}

function removeCommentFromState(commentId: number, self: any) {
	self.updateUnsavedIds(`${commentId}c`, false);
	if (typeof self.issueA11yMessage === "function") {
		self.issueA11yMessage("Comment deleted");
	}
	tagManager({
		action: "Clicked",
		label: "Comment deleted button",
		event: "generic",
		category: "Consultation comments page",
	});
	let comments = self.state.comments;
	const error = "";
	comments = comments.filter(comment => comment.commentId !== commentId);
	self.setState({ comments, error });
	if ((comments.length === 0) && (typeof self.validationHander === "function")) {
		self.validationHander();
	}
}

function removeAnswerFromState(questionId: number, answerId: number, self: any) {
	self.updateUnsavedIds(`${questionId}q`, false);
	let questions = self.state.questions;
	let questionToUpdate = questions.find(question => question.questionId === questionId);
	questionToUpdate.answers = questionToUpdate.answers.filter(answer => answer.answerId !== answerId);
	self.setState({ questions });
	if (typeof self.validationHander === "function") {
		self.validationHander();
	}
	if (typeof self.issueA11yMessage === "function") {
		self.issueA11yMessage("Answer deleted");
	}
	tagManager({
		action: "Clicked",
		label: "Delete answer button",
		event: "generic",
		category: "Consultation comments page",
	});
}

function removeQuestionFromState(questionId: number, self: any) {
	console.log("Called removeQuestionFromState in editing-and-deleting.js");
	console.log(questionId, self);
}
