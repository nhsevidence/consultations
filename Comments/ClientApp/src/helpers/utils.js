import queryString from "query-string";

/**
 * Create a query string out of an object
 * @param obj the key-value pairs that you'd like to be turned into a querystring
 * @returns {string}
 */
export function objectToQueryString(obj) {
	let str = [];
	for (let p in obj)
		if (obj.hasOwnProperty(p)) {
			str.push(encodeURIComponent(p) + "=" + encodeURIComponent(obj[p]));
		}
	if (str.length) return "?" + str.join("&");
	return "";
}

export function queryStringToObject(qryString) {
	return queryString.parse(qryString);
}

/**
 * Resolve a promise on the next tick - used in testing
 * @see https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/#process-nexttick
 * @returns {Promise<any>}
 */
export const nextTick = async () => {
	return new Promise((resolve) => {
		setTimeout(resolve, 0);
	});
};

/**
 * Format a string by replacing {0} with arguments passed in
 * @see https://stackoverflow.com/a/4673436
 * @returns {String}
 */
export function replaceFormat(stringToReplace, args) {
	if (typeof(stringToReplace) === undefined || stringToReplace === null)
		return stringToReplace;

	return stringToReplace.replace(/{(\d+)}/g, function(match, number) {
		return typeof args[number] !== undefined
			? args[number]
			: match
		;
	});
}

/**
 * @param String The string in which you want to check for the presence of "http"
 * Detects link that starts with "http"
 * @returns {Boolean}
 */
export const isHttpLink = link => link.indexOf("http") === 0;
