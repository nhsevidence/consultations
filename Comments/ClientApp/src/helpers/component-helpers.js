// @flow

import React from "react";
import { NavHashLink } from "react-router-hash-link";
import { pullFocusByQuerySelector } from "./accessibility-helpers";

type HashLinkTopType = {
	label: string,
	to: string,
	behavior: string,
	block: string
};

export function HashLinkTop(props: HashLinkTopType) {
	const { label, to, behavior, block } = props;
	return (
		<NavHashLink
			role="menuitem"
			to={to}
			scroll={el => {
				el.scrollIntoView({
					behavior,
					block,
				});
				const element = el.attributes.id.value;
				pullFocusByQuerySelector(`#${element}`);
			}}
		>
			{label}
		</NavHashLink>
	);
}

