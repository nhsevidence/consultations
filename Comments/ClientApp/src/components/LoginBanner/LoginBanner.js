// @flow
import React, { Component, Fragment } from "react";
import { withRouter } from "react-router-dom";
import { DebounceInput } from "react-debounce-input";
import queryString from "query-string";
import Cookies from "js-cookie";
import { appendQueryParameter, removeQueryParameter, removeQuerystring } from "../../helpers/utils";
import { load } from "../../data/loader";

type PropsType = {
	signInURL: string,
	registerURL: string,
	signInButton: boolean,
	signInText?: string,
	match: PropTypes.object.isRequired,
    location: PropTypes.object.isRequired,
	history: PropTypes.object.isRequired,
}

type OrganisationCode = {
	organisationAuthorisationId: number,
	organisationId: number,
	organisationName: string,
	collationCode: string,
}

type StateType = {
	organisationCode: string,
	hasError: bool,
	errorMessage: string,
	showAuthorisationOrganisation: bool,
	authorisationOrganisationFound: OrganisationCode, 
}

export class LoginBanner extends Component<PropsType, StateType> {

	constructor(props: PropsType) {
		super(props);

		this.state = {
			userEnteredCollationCode: "",
			hasError: false,
			errorMessage: "",
			showAuthorisationOrganisation: false,
			authorisationOrganisationFound: null,
		};
	}

	componentDidMount() {
		const userEnteredCollationCode = queryString.parse(this.props.location.search).code;
		if (userEnteredCollationCode){
			this.setState({
				userEnteredCollationCode,
			}, () => {
				this.checkOrganisationCode();
			});
		}
	}

	handleOrganisationCodeChange = (userEnteredCollationCode) => {
		this.setState({
			userEnteredCollationCode,
		}, () => {
			this.checkOrganisationCode(userEnteredCollationCode);
		});
	};

	gatherDataForCheckOrganisationCode = async () => {
		const organisationCode = load(
			"organisation",
			undefined,
			[],
			{
				collationCode: this.state.userEnteredCollationCode,
				consultationId: this.props.match.params.consultationId, 
			})
			.then(response => response.data)
			.catch(err => {
				this.setState({
					hasError: true,
					errorMessage: err.response.data.errorException.Message, 
					showAuthorisationOrganisation: false,					
				});
			});
		return {
			organisationCode: await organisationCode,
		};
	}

	checkOrganisationCode = () => {
		this.gatherDataForCheckOrganisationCode()
			.then(data => {
				if (data.organisationCode != null) {
					this.setState({
						hasError: false,
						errorMessage: "",
						showAuthorisationOrganisation: true,
						authorisationOrganisationFound: data.organisationCode,
					});
				}
			})
			.catch(err => {
				throw new Error("checkOrganisationCode failed " + err);
			});		
	}

	gatherDataForCreateOrganisationUserSession = async () => {
		const sessionId = load(
			"organisationsession",
			undefined,
			[],
			{
				collationCode: this.state.userEnteredCollationCode,
				organisationAuthorisationId: this.state.authorisationOrganisationFound.organisationAuthorisationId, 
			}, 
			"POST")
			.then(response => response.data)
			.catch(err => {
				this.setState({
					hasError: true,
					errorMessage: err.response.data.errorException.Message, 
					showAuthorisationOrganisation: false,
				});
			});
		return {
			sessionId: await sessionId,
		};
	}

	CreateOrganisationUserSession = async () => {
		const sessionId = this.gatherDataForCreateOrganisationUserSession()
			.then(data => {
				if (data.sessionId != null) {
					this.setState({
						hasError: false,
						errorMessage: "",						
					});
					return data.sessionId;
				}
			})
			.catch(err => {
				throw new Error("checkOrganisationCode failed " + err);
			});		
		return {
			sessionId: await sessionId,
		};
	}

	handleConfirmClick = () => {
		var consultationId = this.props.match.params.consultationId;
		this.CreateOrganisationUserSession().then(data => {
			Cookies.set(`ConsultationSession-${consultationId}`, data.sessionId); //TODO: add to cookie policy !!!!!!!!!!!!!!!!!!!!!!!!!!!!!

			//now, set state to show logged in. 
			this.setState({
				showAuthorisationOrganisation: false,
				//TODO: new property here for hiding the org code entry.
			})

			//TODO: then update the Context, to recognise the cookie.

		})
		.catch(err => {
			this.setState({
				hasError: true,
				errorMessage: "Unable to confirm",
			})
		});			
	}


	render(){
		const {userEnteredCollationCode} = this.state;
		const { match, location, history } = this.props

		return (
			<div className="panel panel--inverse mt--0 mb--0 sign-in-banner"
					 data-qa-sel="sign-in-banner">
				<div className="container">
					<div className="grid">
						<div data-g="12">
							<div className="LoginBanner">
								<p>If you would like to comment on this consultation as part of an organisation, please enter your organisation code here:</p>
								<label>
									Organisation code 
									{this.state.hasError && 
										<div>{this.state.errorMessage}</div>
									}
									<DebounceInput
										minLength={6}
										debounceTimeout={400}
										type="text"
										onChange={e => this.handleOrganisationCodeChange(e.target.value)}
										className="form__input form__input--text"
										data-qa-sel="OrganisationCodeLogin"
										id="collationCode"
										value={userEnteredCollationCode}
									/>
								</label>
								<br/><br/>
								{this.state.showAuthorisationOrganisation && 
									<Fragment>
										<label>Confirm organisation name
											<strong>{this.state.authorisationOrganisationFound.organisationName}</strong>
										</label>
										<br/>
										<button className="btn btn--cta" onClick={() => this.handleConfirmClick()}  title={"Confirm your organisation is " + this.state.authorisationOrganisationFound.organisationName}>Confirm</button>
									</Fragment>
								}								
								<a href={this.props.signInURL} title="Sign in to your NICE account">
									Sign in to your NICE account</a> {this.props.signInText || "to comment on this consultation"}.{" "}
								<br/>
								Don't have an account?{" "}
								<a href={this.props.registerURL} title="Register for a NICE account">
									Register
								</a>
							</div>
							{this.props.signInButton &&
								<p>
									<a className="btn btn--inverse" href={this.props.signInURL} title="Sign in to your NICE account">Sign in</a>
								</p>
							}
						</div>
					</div>
				</div>
			</div>
		);
	}
}

export default withRouter(LoginBanner); 