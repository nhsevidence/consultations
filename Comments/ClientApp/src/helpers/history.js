import { canUseDOM } from "./utils";

let browserHistory = null;

// Browser history only makes sense on the client
if(canUseDOM()) {
	const createHistory = require("history/createBrowserHistory").default;
	browserHistory = createHistory();
}

export default browserHistory;
