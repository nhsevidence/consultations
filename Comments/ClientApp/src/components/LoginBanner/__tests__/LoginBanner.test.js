/* global jest */

import React from "react";
import { mount } from "enzyme";
import { MemoryRouter } from "react-router";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import toJson from "enzyme-to-json";

import { nextTick } from "../../../helpers/utils";
import OrganisationCode from "./OrganisationCode.json";
import { LoginBanner } from "../LoginBanner";

const mock = new MockAdapter(axios);

jest.mock("../../../context/UserContext", () => {
	return {
		UserContext: {
			Consumer: (props) => {
				return props.children({
					isAuthorised: true,
				});
			},
		},
	};
});

describe("[ClientApp] ", () => {
	describe("LoginBanner Component", () => {
		const fakeProps = {
			match: {
				url: "/1/1/introduction",
				params: {
					consultationId: 1,
				},
			},
			location: {
				pathname: "/1/1/introduction",
				search: "",
			},
			history:{
				location:{
					search: "",
				},
				listen: function(){},
			},
			signInURL: "/signin-url",
			registerURL: "/register-url",
			signInButton: true,
			signInText: "signin text",
		};

		beforeEach(() => {
			mock
				.onGet()
				.reply(200, OrganisationCode);
		});

		afterEach(() => {
			window.__PRELOADED__ = { isAuthorised: true};
			mock.reset();
		});

		it("should match snapshot. no querystring", async () => {
				 
			const wrapper = mount(				
				<LoginBanner {...fakeProps} />				
			);
			
			await nextTick();
			wrapper.update();

			expect(toJson(wrapper, {noKey: true, mode: "deep",})).toMatchSnapshot();

			wrapper.unmount();
		});

		it("organisation code should be prepopulated from querystring", async () => {

			const expectedOrganisationCode = "123412341234";
			var updatedProps = Object.assign(fakeProps, {location: {pathname: "/1/1/introduction", search:`?code=${expectedOrganisationCode}`}})

			const wrapper = mount(				
				<LoginBanner {...updatedProps} />				
			);
			
			await nextTick();
			wrapper.update();

			expect(wrapper.state().userEnteredCollationCode).toEqual(expectedOrganisationCode);

			expect(toJson(wrapper, {noKey: true, mode: "deep",})).toMatchSnapshot();

			wrapper.unmount();
		});

	});
});