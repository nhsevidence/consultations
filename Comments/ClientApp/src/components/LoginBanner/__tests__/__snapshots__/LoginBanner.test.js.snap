// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ClientApp]  LoginBanner Component should match snapshot. no querystring 1`] = `
<div
  className="panel panel--inverse mt--0 mb--0 sign-in-banner"
  data-qa-sel="sign-in-banner"
>
  <div
    className="container"
  >
    <div
      className="grid"
    >
      <div
        data-g="12"
      >
        <div
          className="LoginBanner"
        >
          <a
            href="/signin-url"
            title="Sign in to your NICE account"
          >
            Sign in to your NICE account
          </a>
           
          signin text
          .
           
          <br />
          Don't have an account?
           
          <a
            href="/register-url"
            title="Register for a NICE account"
          >
            Register
          </a>
        </div>
        <p>
          <a
            className="btn btn--inverse"
            href="/signin-url"
            title="Sign in to your NICE account"
          >
            Sign in
          </a>
        </p>
      </div>
    </div>
  </div>
</div>
`;
